src/Sulu/Bundle/MediaBundle/DependencyInjection/StorageCompilerPass.php

```diff
@@ -0,0 +1,221 @@
+            [
+                's3' => [
+                    'class' => 'Sulu\Bundle\MediaBundle\Media\Storage\Resolver\Flysystem\AwsS3Resolver',
+                    'tag' => 'League\Flysystem\AwsS3v3\AwsS3Adapter',
+                ],
+                'dropbox' => [
+                    'class' => 'Sulu\Bundle\MediaBundle\Media\Storage\Resolver\Flysystem\DropBoxResolver',
+                    'tag' => 'League\Flysystem\AwsS3v3\DropboxAdapter',
+                ],
+            ]
+        );
+
+        // get all available adapters
+        $taggedServices = $container->findTaggedServiceIds('sulu_media.storage_adapter');
+        $this->adapters = [];
+        foreach ($taggedServices as $id => $tags) {
+            foreach ($tags as $attributes) {
+                $this->adapters[$attributes['alias']] = $id;
+            }
+        }
+
+        // use only configured adapters
+        $storageAdapters = $container->getParameter('sulu_media.storage.adapters');
+        $storageManagerDefinition = $container->getDefinition('sulu_media.storage_manager');
+        foreach ($storageAdapters as $alias => $config) {
+            // create new storage definition
+            $id = $this->getStorageDefinition($container, $alias, $config);
+
+            // add storage to manager
+            $storageManagerDefinition->addMethodCall(
+                'add',
+                [new Reference($id), $alias]
+            );
+        }
+    }
+
+    /**
+     * @param ContainerBuilder $container
+     * @param $alias
+     * @param $config
+     *
+     * @return string
+     *
+     * @throws StorageAdapterNotFoundException
+     */
+    protected function getStorageDefinition(
+        ContainerBuilder $container,
+        $alias,
+        $config
+    ) {
+        if (!$config['type']) {
+            throw new StorageAdapterNotFoundException(sprintf('Storage adapter for "%s" not found!', $alias));
+        }
+
+        // get abstract storage
+        $adapterName = $this->getAdapter($config['type']);
+        unset($config['type']);
+        $id = sprintf('sulu_media.%s_storage', $alias);
+
+        // create definition by abstract adapters
+        $storageDefinition = $container->setDefinition($id, new DefinitionDecorator($adapterName));
+
+        // get adapter classname
+        $className = $container->getParameterBag()->resolveValue(
+            $container->getDefinition($adapterName)->getClass()
+        );
+
+        // get reflection class to set constructor correct
+        try {
+            $class = new \ReflectionClass($className);
+        } catch (\ReflectionException $e) {
+            throw new StorageAdapterNotFoundException(
+                sprintf(
+                    'Class Reflection error for "%s": ' . $e->getMessage() . PHP_EOL .
+                    $e->getTraceAsString(), $className
+                ));
+        }
+
+        // set the new service constructor
```

Comment:
you can create e.g. 2 local storages with following config:

```
local:
    type: local
    uploadPath: %kernel.root_dir%/../uploads/media
local2:
    type: local
    uploadPath: %kernel.root_dir%/../uploads/media2
```

this will create 2 new services which extend the `sulu_media.storage.adapter.local`.

it use the construct parameters to set it correctly for all types. In this case the construct function has a $uploadPath parameter. the reflection is used to know which parameter this is and will replace it. 

for the flysystem type the constructor is another but it still works the same way and you can use a storage adapter of the same type multiple with other configuration.